### ISSUE TITLE
Issue parsing unknown enum case

### ISSUE THREAD

**Issue by mpdifran** [apple/swift-protobuf#972]
iOS 13.4
Xcode Version 11.4 (11E146)
Swift 5
Swift-Protobuf 1.8.0
proto2

My iOS app is using generated code from outdated proto files. Since I've compiled the `.swift` files from the old protos, we've added some new cases to one of our enums, and started returning the cases to the client from the backend. I'm getting an error when parsing the JSON data into the Swift Proto objects: `unrecognizedEnumValue`.

From my understanding of the proto spec, it should be fine to add new enum cases, and clients who don't have the latest spec map the unknown value to the 0 case. Is there something wrong with how we name our 0th case?

Old Proto
```
message Product {
    enum State {
        State_INVALID = 0;
        ACTIVE = 1;
        DISCONTINUED = 2;
    }
    optional State state = 1;
}
```

New Proto
```
message Product {
    enum State {
        State_INVALID = 0;
        ACTIVE = 1;
        DISCONTINUED = 2;
        FOR_SALE = 3;
        NOT_FOR_SALE = 4;
    }
    optional State state = 1;
}
```

JSON Being Parsed
```
{
    "state": "FOR_SALE"
}
```

Swift JSON Decoding (using Old Proto)
```
do {
    var options = JSONDecodingOptions()
    options.ignoreUnknownFields = true
    return try Product(jsonUTF8Data: data, options: options)
} catch {
    print(error) // error is printed unrecognizedEnumValue
    return nil
}
```

**Comment by tbkka** [comment]
> From my understanding of the proto spec, it should be fine to add new enum cases, and clients who don't have the latest spec map the unknown value to the 0 case.

Has this changed recently?  Last I checked, this was not in fact the behavior of Google's protobuf implementations for other languages.

As a rule, we try to stick very closely to the features supported by the main protobuf project in order to avoid confusion and problems for people that are mixing multiple protobuf implementations.  Please file an issue with the protocolbuffers/protobuf project and let us know what happens.  If you can convince them to adopt this behavior, we'll happily follow suit.

**Comment by tbkka** [comment]
One complicating factor here is that the binary decoder is able to preserve the unsupported value so that later re-serialization will preserve it even if it's unknown.  The JSON decoder cannot do this, at least it cannot do it _for textual field values_.

However, there is an option for encoding enums into JSON documents as numeric values.  With that, your example above would be
```
{
   "state": 3
}
```
It would be possible in theory for the decoder to preserve the unsupported numeric value in the same way the binary decoder does.  We don't do this today, but it might be worth discussing with the protocolbuffers/protobuf folks as a possible option.  As above, it would be a lot easier for us to justify something like that if they adopted it first.

**Comment by thomasvl** [comment]
First up - one confusing thing in the docs is last time I checked, there isn't always a clear distinction between evolving the proto when using binary vs. json encodings.  Enums can be added to using the binary encoding, but not so much using the JSON encoding, because there is no way for a client to know what the string should mean.  @tbkka suggestion of using the `alwaysPrintEnumsAsInts` encoding option might make this better, since it would be raw numbers and proto3 syntax has better handling of decoding things like that.

The `ignoreUnknownFields` decoding option is about fields, but in your case the field (_state_), is known, it is just a string value that is unknown, and that's why it errors.

**Comment by mpdifran** [comment]
Thanks for the information! 

I forgot to include it in my original comment, but we're using `proto2`. I've updated it to reflect that.

We've done some research on this, and we've found the following:

#### Proto2 Unrecognized Enum Cases
> Notably, unrecognized enum values are discarded when the message is deserialized, which makes the field's has.. accessor return false and its getter return the first value listed in the enum definition, or the default value if one is specified.

Reference: https://developers.google.com/protocol-buffers/docs/proto#updating

#### Proto3 Default Values
> For enums, the default value is the first defined enum value, which must be 0.

Reference: https://developers.google.com/protocol-buffers/docs/proto3#default

It looks like there's already a PR open to add this feature, is it possible to merge this into master?

https://github.com/apple/swift-protobuf/pull/896

I'll look into opening a similar issue on `protocolbuffers/protobuf` as well!

**Comment by thomasvl** [comment]
The docs you quote are talking about the binary encoding, not the JSON one.  That's one of the cases where they don't make a clear distinction between encodings that I mentioned.

As for #896, as the comment suggest, we're hesitant to add the feature because the main protobuf library doesn't have it for any other language.  If they add then support we'd be willing to added support to the Swift version to match.

**Comment by mpdifran** [comment]
Ah ok, thanks for clarifying! I'll see if decoding from binary handles this better.

**Comment by thomasvl** [comment]
`Test_JSON.swift` - `testOptionalNestedEnum` does test `{"optionalNestedEnum":123}` <=> `.UNRECOGNIZED(123)` round tripping for proto3 syntax. So using the option `alwaysPrintEnumsAsInts` when encoding should give you something that should work (again with proto3 syntax).

If you are in proto2 syntax, I don't immediately see a test case that would cover an unknown numeric (and stuff it into the unknown fields), but I'd also have to try things on the C++ library to confirm what they do in that case also.

**Comment by srcreigh** [comment]
@thomasvl Thanks for all your time spent on this. The documentation on this repo says that unrecognized JSON enum values are ignored, is this an issue with the documentation?

> If deserialization encounters an unknown value, that value is either ignored (JSON)
> *in the Enum API section [here](https://github.com/apple/swift-protobuf/blob/master/Documentation/API.md)*

**Comment by thomasvl** [comment]
Yup, documentation error.

**Comment by srcreigh** [comment]
It may also be worthwhile to put some kind of notice at the top of the README. I understand there is a technical reason to not support forwards compatibility in JSON-encoded enums, but I think lots of people have come to expect forwards compatibility from protobuf libraries (whatever the serialization mechanism). Especially when it comes to swift on iOS, it could take a while to deploy fixes to buggy clients, especially if the app isn't in control of every data source. In short I think it would be a net win to advise people to avoid JSON serialization and proto2 if forwards compatibility isn't supported under those configurations

**Comment by thomasvl** [comment]
Doesn't that comment more apply to protobuf and not just the Swift implementation?  The behavior (and defaults and options) we added all model what the main library does/provides.  It isn't like we've chosen something different here.

**Comment by srcreigh** [comment]
I understand how difficult it can be push things through and get answers internally at google about protobuf related things, trust me! I have imagined that such an advertisement here could even generate enough demand to push this kind of change through to the main library. I don't imagine that there are many teams at Google who rely on JSON serialization of protocol buffers or even the swift language itself, so it makes sense to me that it wouldn't be a priority for them to fix this (maybe ever).

In the meantime this library (whatever its dependencies) does not generate forwards compatible code so we can't use it for iOS apps. I think we are experimenting with a forked version of this library which ignores unrecognized enums. Practically we need some way to ship a forwards compatible iOS app. I appreciate your sharing some solutions re; ignoring enums on the server side. Thankfully that works for us, but it wouldn't work for any iOS apps that interface with external APIs that could send unrecognized enums. I would appreciate any updates about this lib generating forwards compatible JSON code, but you have already done so much work in this lib I couldn't ask for more. Thanks again.

**Comment by thomasvl** [comment]
> In the meantime this library (whatever its dependencies) does not generate forwards compatible code so we can't use it for iOS apps.

Can you please explain this comment?  The code is working to the protobuf spec.  If you use the option to always serialize enums as integers (which isn't the protobuf spec behavior but is documented as an option) it does improve things in this specific case.

It also sounds like you might be asking about interop with other services that doen't use protobuf?  Not all JSON is the same, the protobuf spec includes a section about how things should be converted to/form JSON.  All libraries that claim to follow the protobuf spec work the same way.  The protobuf repository includes a bunch of JSON specific conformance tests to help with ensure they all act the same way.

Silently ignoring something the library doesn't understand also does not make things "forward compatible".  Yes, it means you can skip the data, but depending on what it was, it can fundamentally change the meaning of things.  If you were to take in something, drop the thing you don't understand, and then send it back to the server; you could end up changing that value on the server, which could have implications for the service/information.

**Comment by srcreigh** [comment]
For example, if a new enum field is shipped on a proto with v12.5 of an iOS app but is totally ignored (maybe it's a new feature under a shared schema that is implemented on a web app and is not implemented on the mobile app for a while). Then when this field, which is totally ignored, gets a new enum value, the v12.5 version of the app breaks because the only thing this that swift-protobuf can do is throw an exception when it sees the unrecognized enum value.

Of course this scenario is the core problem that protocol buffers as a whole set out to solve.

> Protocol buffers were designed to solve many of these problems:
> - New fields could be easily introduced, and intermediate servers that didn't need to inspect the data could simply parse it and pass through the data without needing to know about all the fields.
> https://developers.google.com/protocol-buffers/docs/overview#a-bit-of-history

(Of course an iOS app is not a server but it's the same thing as far as this is concerned because it participates in a distributed protocol.)

I'm not suggesting that this lib doesn't meet the spec you set out for yourselves, but something definitely got missed in the process of drafting and implementing protocol buffers spec if it doesn't meet the fundamental design criteria of supporting forwards compatibility in distributed protocols.

**Comment by tbkka** [comment]
Unfortunately, the SwiftProtobuf project is just implementing the spec.  The spec itself is the responsibility of the protocolbuffers project:
https://github.com/protocolbuffers/protobuf
Please file an issue there and let us know how that discussion goes.

**Comment by tbkka** [comment]
According to @acozzette, the option to drop unknown fields also applies to unrecognized enum cases.  I've asked for further clarification for the case of repeated enum fields.

**Comment by tbkka** [comment]
@srcreigh or @mpdifran -- If you could send us a PR with the necessary change, we'd greatly appreciate.  The code change looks trivial; just modify the Enum methods in JSONDecoder.swift to ignore Enum decoding errors when `options.ignoreUnknownFields` is set.

The PR should also include some test cases.  You can probably adjust `SwiftProtobufTests/Test_JSONDecodingOptions.swift` to use something other than `Empty` as the test message (pretty much any message with valid enum fields) to verify that an unknown enum case in a valid field is ignored when that flag is set.

**Comment by mpdifran** [comment]
Can do! 

There is #896 which is currently open, but I can make a new PR with the changes you suggested if you prefer!

**Comment by tbkka** [comment]
A-ha!  Thank you for reminding me.  I've requested changes to that PR to align it with the behavior described by @acozzette.  Then I think we're good to go.

**Comment by loganchang** [comment]
Hello all! Are there any updates for this issue and #896? I've currently encountered this same error and would love to see this merged in.

**Comment by thomasvl** [comment]
We're still hoping for some guidance from the core protocolbuffers library for the behaviors, we don't want to accidentally do something slight different in the edge cases.

**Comment by loganchang** [comment]
Sounds good, will comment on that issue as well! Thanks!

**Comment by mpdifran** [comment]
Any ETA on when we can get this into an official release? Is there anything I can do to help move this along? We've been using this branch on our fork for a while and we'd like to get back to the main releases.

**Comment by cassianomonteiro** [comment]
@thomasvl Looks like it's moving forward as per https://github.com/protocolbuffers/protobuf/pull/9534, could you please help us getting this merged into the swift library?

**Comment by thomasvl** [comment]
I can't commit at time in the near future to this specifically.  Once the conformance test lands, we might have to add an expected failure if the library can't currently support the new test.  Someone else can always look at what it would take to support this and propose a PR for the changes.

**Comment by tbkka** [comment]
To elaborate on what Thomas said: We really want to wait for protocolbuffers/protobuf#9534 to actually get merged, since it will provide the test that we need to ensure that #896 really does match the revised specification.

**Comment by thomasvl** [comment]
Looks like protocolbuffers/protobuf#9534 is getting merged soon, as far as a flag for controlling the behavior it will be reusing the unknown field one. Once it's merged we can look to update.
